### DISCUSSION TITLE
Unsupported color conversion request with DCM files.

### DISCUSSION THREAD

**Discussion by inputfalken** [dlemstra/Magick.NET#1738] [General] [answered]
Hi not sure if this a bug but I figured i would start with a question here. 

I'm trying to convert dicom files whose image compression is JPEG lossless compression.  I have tried both with this library and the
imagemagick CLI tool and get the same error.

Like the following  example `magick.exe foo.dcm bar.jpg` -> `Unsupported color conversion request @ error/jpeg.c/JPEGErrorHandler/356.` 
When I run the verbose flag i get the error the moment it want to process the pixel data.

My theory is that it's due to the following:
> Lossless RGB JPEG files
This requires the SmartScale feature, which is not supported.
 Source: https://github.com/ImageMagick/jpeg-turbo?tab=readme-ov-file#not-supported
 
 I'm wondering what my options are.  I would love to be able to continue using your .NET solution of imagemagick. ❤️

**Comment by dlemstra** [comment]
Can you share a `dcm` file that reproduces this so I can check with a debugger what is happening.

**Reply by inputfalken** [reply]
Thank you for looking into this!  I have emailed you a anonymized sample file with this particular issue. 😊

**Reply by dlemstra** [reply]
Thanks for sharing the file. It looks like the embedded jpeg file is stored in a format that is not supported by the library that is being used to read the file. The colorspace is `YCbCr` and `lossless` and `libjpeg-turbo` has no support for reading this. It might also be possible that the embedded jpeg file is written incorrectly.

**Reply by inputfalken** [reply]
Thanks for the response! 😃

I have been able to open & display the image in GUI applications such as MicroDicom. 

Do you have any suggestion on a solution that could make it possible to process these images (Ideally with ImageMagick & this library)?

**Reply by dlemstra** [reply]
I suspect that viewer contains a hack to make it possible to read that embedded jpeg file. I played around with a debugger and marked the file as non lossless. I then get an output image that has purple colors. That is probably happening because the colorspace is incorrect in the file. Changing it to `RGB` fixes the problem with this image. I think you will need to reach out to the vendor that created those file and let them know they are creating a corrupt jpeg file.

**Reply by inputfalken** [reply]
That could very well be the case!

Is there anyway that I could force `Magick.NET` or `Imagemagick` to treat the image as non lossless and also change the colospace to be for instance `grayscale`? 

Thanks for the thorough investigation of my problem! I will provide the information you have provided forward 👍

**Reply by dlemstra** [reply]
I did some more research and it seems that `libjpeg-turbo` determines that the colorspace of the image is `YCbCr` because the image contained a `JFIF` marker. But I wonder if this should not always be `RGB` when the image is `lossless`. We are unable to add a patch for that in ImageMagick because we cannot check if the image is `lossless`. The field that contains this value is private in the `libjpeg-turbo` api.

I would like to open an issue in the `libjpeg-turbo` project to ask for information about this. But they probably want to see the jpeg file to check the issue. Would you be okay with me sharing the embedded jpeg file from the dicom file? I can email you the file that I want to share before sharing it in an issue in that project.

I do have a nasty hack that you could use when you run into this exception and still want to read the file.

```C#
private void BreakJFIFHeaderForLosslessCompressedFile(ref Span<byte> bytes)
{
    for (var i = 0; i < bytes.Length - 10; i++)
    {
        // Search for jpeg and JFIF header
        if (bytes[i] == 0xff && bytes[i + 1] == 0xd8 && bytes[i + 2] == 0xff &&
            bytes[i + 6] == 'J' && bytes[i + 7] == 'F' && bytes[i + 8] == 'I' && bytes[i + 9] == 'F')
        {
            // Break the JFIF header by changing it to JFIX
            bytes[i + 9] = (byte)'X';
            break;
        }
    }
}
```

This only breaks the JFIF header and does not check if the image was compressed lossless so you would need to read the image first and wait for that exception message. Then call this method and read the image again.

**Reply by inputfalken** [reply]
Sounds good to share the embedded JPEG file with `libjpeg-turbo`. The image should not have anything notable in it. 👍 
Would be amazing if that would solve/support these kinds of images. 

If you don't mind, please share a link to the `libjpeg-turbo` issue when you have created it.  I'm quite curious to follow it.😃

**Reply by dlemstra** [reply]
I have just created this issue: https://github.com/libjpeg-turbo/libjpeg-turbo/issues/793

**Reply by dlemstra** [reply]
We got a response and it looks like this should be fixed in the library that wrote this file.  I don't think we should add a patch for this in ImageMagick because this seems to be an edge case.

**Reply by inputfalken** [reply]
That's fantastic to hear! Thank you for the time and effort you put into my question.

Sounds good to wait for `libjpeg-turbo` to fix it.  👍

But the plan then is to wait for `libjpeg-turbo` to perform the fix. Followed by a new version of ImageMagick where the fix has been applied? If so, would it be considered a patch version for ImageMagick to use the updated `libjpeg-turbo`?  I'm mainly wondering if this can be released fairly shortly.

**Reply by dlemstra** [reply, marked as answer]
I think you need to read me answer and the referenced issue again :smile:.  "the library that wrote this file" means the vendor that created this dicom file.

**Reply by inputfalken** [reply]
My bad, I thought you ment `libjpeg-turbo` was the library.  😅

But i see now you said **library that wrote**, I will provide this information to the vendor. 👍

**Reply by dlemstra** [reply]
Make sure you keep following the conversation in the issue that I opened it contains some tips for LEADTOOLS.

**Reply by inputfalken** [reply]
I have been looking over it during the last days. 🙂

I'm planning to contact them and reference the issue you created.

**Reply by inputfalken** [reply]
Hi again, I just wanted to inform you that after a long conversation where I referenced both this and the [libjpeg/turbo issue](https://github.com/libjpeg-turbo/libjpeg-turbo/issues/793). 

They have concluded that everything looks fine on their end. And they will not do anything with their lossless JPEG implementation.

**Reply by dlemstra** [reply]
What an interesting conclusion.... I suspect that you will then need to use my `BreakJFIFHeaderForLosslessCompressedFile` method to make it possible to read the image.

**Reply by inputfalken** [reply]
It looks that way. 

Thank you for your thorough investigation and for providing a workaround for this issue.
